File: skills/plotting/python/pavlab_heatmap.py
"""pavlab_heatmap (Python) — opinionated seaborn.heatmap wrapper.

Mirror of R/pavlab_heatmap.R. Same lab defaults, same five named divergent
palettes, same auto-behavior for correlation/expression/raw modes. Outputs
are publication-grade matplotlib figures.

Usage:
    from pavlab_heatmap import pavlab_heatmap

    pavlab_heatmap(expr_df)                                  # expression mode (default)
    pavlab_heatmap(expr_df, filename="fig2a.pdf")            # save to file
    pavlab_heatmap(cor_df, mode="correlation")               # correlation mode
    pavlab_heatmap(raw_df, mode="raw")                       # raw mode
"""

from __future__ import annotations

from pathlib import Path
from typing import Any, Literal

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

try:
    from .palettes import black_body_palette, divergent_palette  # package-style
except ImportError:
    from palettes import black_body_palette, divergent_palette  # script-style (sys.path)

Mode = Literal["expression", "correlation", "raw"]


def pavlab_heatmap(
    mat,
    mode: Mode = "expression",
    cluster_rows: bool = False,
    cluster_cols: bool = False,
    show_rownames: bool | None = None,
    show_colnames: bool | None = None,
    standardize_rows: bool | None = None,
    zclip: float | None = 3.0,
    divergent: bool | None = None,
    na_color: str = "#cccccc",
    palette_n: int = 256,
    rowname_threshold: int = 50,
    colname_threshold: int = 30,
    angle_col: int = 45,
    border_color: str | None = None,           # None = no gridlines
    correlation_negative_tolerance: float = -0.1,
    filename: str | Path | None = None,
    figsize: tuple[float, float] | None = None,
    cmap=None,
    ax=None,
    **kwargs: Any,
):
    """Pavlab-style heatmap (Python port of R pavlab_heatmap).

    Wraps seaborn.heatmap with lab defaults. Any unrecognized kwarg is
    forwarded to seaborn.heatmap. See ../SKILL.md for the rationale behind
    each default.

    Parameters
    ----------
    mat : 2D array, pandas DataFrame, or anything convertible.
        The matrix to plot. If a DataFrame, index/columns become labels.
    mode : {"expression", "correlation", "raw"}
        Sets transformation and palette defaults.
    cluster_rows, cluster_cols : bool
        Default False. (seaborn doesn't cluster in `heatmap`; use
        `sns.clustermap` if you want clustering — this wrapper doesn't
        switch backends for you, but will warn if you ask for clustering.)
    standardize_rows : bool, optional
        None = True iff mode == "expression". Computes row Z-scores.
    zclip : float, optional
        Clip standardized values to [-zclip, +zclip]. Default 3.
    divergent : bool, optional
        Force divergent vs sequential palette. None = auto from mode +
        data. In correlation mode, if all values are >= correlation_negative_tolerance,
        small negatives are clipped to 0 and the sequential black-body
        palette is used.
    na_color : str
        Color for NaN cells. Default "#cccccc" (matches R grey80).
    border_color : str or None
        Cell-edge color (forwarded to seaborn as `linecolor`). Default
        None = no gridlines (sets linewidths=0).
    correlation_negative_tolerance : float
        See SKILL.md. Default -0.1.
    filename : str | Path | None
        If provided, the figure is saved via fig.savefig and not shown.
    figsize : (w, h) tuple, optional
        Override the default figure size.
    cmap : matplotlib Colormap, optional
        Override the auto-chosen palette.
    ax : matplotlib Axes, optional
        Draw into an existing Axes (figsize and filename are ignored).
    **kwargs
        Forwarded to seaborn.heatmap.

    Returns
    -------
    matplotlib Axes object containing the heatmap.
    """
    # --- normalize input ---
    if isinstance(mat, pd.DataFrame):
        mat = mat.copy()
        is_df = True
    else:
        mat = np.asarray(mat, dtype=float)
        mat = pd.DataFrame(mat)
        is_df = False  # noqa: F841

    if cluster_rows or cluster_cols:
        # Pheatmap/seaborn split: clustering needs clustermap, not heatmap.
        # We don't transparently switch; surface this to the caller.
        raise NotImplementedError(
            "pavlab_heatmap (Python) does not currently dispatch to "
            "sns.clustermap. Call sns.clustermap directly if you need "
            "clustering; the same palette helpers work via cmap=."
        )

    # --- resolve auto-defaults ---
    if standardize_rows is None:
        standardize_rows = (mode == "expression")
    if show_rownames is None:
        show_rownames = mat.shape[0] <= rowname_threshold
    if show_colnames is None:
        show_colnames = mat.shape[1] <= colname_threshold

    # --- mode-specific transforms ---
    if mode == "correlation":
        if mat.shape[0] == mat.shape[1]:
            arr = mat.to_numpy(copy=True)
            np.fill_diagonal(arr, np.nan)
            mat = pd.DataFrame(arr, index=mat.index, columns=mat.columns)
        else:
            import warnings
            warnings.warn(
                "pavlab_heatmap: mode='correlation' but matrix isn't square; "
                "leaving diagonal alone"
            )

    if standardize_rows:
        # Row Z-score: (x - row mean) / row std.
        row_mean = mat.mean(axis=1)
        row_std = mat.std(axis=1, ddof=1).replace(0, np.nan)
        mat = mat.sub(row_mean, axis=0).div(row_std, axis=0)
        if zclip is not None and np.isfinite(zclip) and zclip > 0:
            mat = mat.clip(lower=-zclip, upper=zclip)

    # --- palette decision (mirrors R logic) ---
    if cmap is None:
        if divergent is None:
            if mode == "expression":
                divergent = True
            elif mode == "correlation":
                min_val = np.nanmin(mat.to_numpy())
                if not np.isfinite(min_val):
                    divergent = False
                elif min_val >= correlation_negative_tolerance:
                    if min_val < 0:
                        # Clip small negatives to 0 (the cutpoint).
                        mat = mat.clip(lower=0)
                    divergent = False
                else:
                    divergent = True
            else:  # raw
                vals = mat.to_numpy()
                vals = vals[np.isfinite(vals)]
                divergent = vals.size > 0 and (vals < 0).any() and (vals > 0).any()

        if divergent:
            cmap = divergent_palette(palette_n)
        else:
            cmap = black_body_palette(palette_n)
        cmap.set_bad(na_color)

    # --- breaks: symmetric for divergent ---
    vmin = vmax = None
    if divergent:
        max_abs = np.nanmax(np.abs(mat.to_numpy()))
        if not np.isfinite(max_abs) or max_abs == 0:
            max_abs = 1.0
        vmin, vmax = -max_abs, max_abs

    # --- figure setup ---
    own_fig = False
    if ax is None:
        if figsize is None:
            # mild auto-sizing
            figsize = (max(5, mat.shape[1] * 0.25 + 2),
                       max(4, mat.shape[0] * 0.22 + 1))
        fig, ax = plt.subplots(figsize=figsize)
        own_fig = True
    else:
        fig = ax.figure

    # --- forward to seaborn ---
    linewidths = 0 if border_color is None else 0.5
    linecolor = border_color if border_color is not None else "white"

    sns_kwargs = dict(
        cmap=cmap,
        vmin=vmin, vmax=vmax,
        xticklabels=show_colnames,
        yticklabels=show_rownames,
        linewidths=linewidths,
        linecolor=linecolor,
        cbar=True,
        square=False,
        ax=ax,
    )
    sns_kwargs.update(kwargs)
    sns.heatmap(mat, **sns_kwargs)

    # Rotate column labels (the R angle_col equivalent).
    if show_colnames:
        plt.setp(ax.get_xticklabels(), rotation=angle_col,
                 ha="right", rotation_mode="anchor")

    if own_fig:
        fig.tight_layout()
        if filename is not None:
            fig.savefig(filename, dpi=200)
            plt.close(fig)

    return ax

File: skills/plotting/R/pavlab_heatmap.R
# pavlab_heatmap: opinionated pheatmap wrapper encoding the lab's default
# choices for publication-grade heatmaps. See ../SKILL.md for the rationale
# behind each default.
#
# Source palettes.R alongside this file:
#   source("R/palettes.R"); source("R/pavlab_heatmap.R")
#
# Dependencies: pheatmap (>= 1.0.12), grDevices, stats (all stdlib in R >= 4).

#' Pavlab-style heatmap.
#'
#' Wraps pheatmap::pheatmap with lab-standard defaults: no clustering,
#' grey for missing values, row-standardization for expression mode,
#' Z-clipping at +/- 3, divergent blue-black-yellow palette for signed data,
#' black-body sequential palette for non-negative data, NA on correlation
#' diagonals, and automatic row/column-label hiding when there are too many
#' labels to be readable.
#'
#' Any argument unrecognized by this function is forwarded to pheatmap().
#'
#' @param mat Numeric matrix to plot.
#' @param mode One of "expression", "correlation", "raw". Determines
#'   transformation and palette defaults; see SKILL.md.
#' @param cluster_rows,cluster_cols Logical. Default FALSE.
#' @param show_rownames,show_colnames Logical or NULL (auto).
#' @param standardize_rows Logical or NULL (NULL = TRUE iff mode is
#'   "expression").
#' @param zclip Numeric. Z-score values are clipped to [-zclip, +zclip] when
#'   standardize_rows is TRUE. Default 3.
#' @param divergent Logical or NULL (auto). NULL: TRUE for correlation or
#'   expression, OR for raw data that spans zero.
#' @param na_color Color for missing values. Default "grey80".
#' @param palette_n Number of palette colors. Default 256.
#' @param rowname_threshold,colname_threshold Integer. Auto-hide labels when
#'   the corresponding dimension exceeds the threshold. Defaults 50 (rows)
#'   and 30 (cols).
#' @param angle_col Column-label rotation. Default 45.
#' @param border_color Cell border color. Default NA (no gridlines between
#'   cells — pheatmap's default `grey60` gridlines visually clutter the
#'   color signal).
#' @param correlation_negative_tolerance In `mode = "correlation"`, values
#'   between this threshold and 0 are treated as "effectively zero" — clipped
#'   to 0 and the palette stays sequential (black-body). Only when there are
#'   substantial negative correlations (below this threshold) do we fall back
#'   to the divergent palette. Default -0.1.
#' @param filename If non-NULL, written via pheatmap's filename arg
#'   (auto-detects format from extension).
#' @param ... Forwarded to pheatmap::pheatmap.
#'
#' @return The pheatmap object (invisibly), as returned by pheatmap::pheatmap.
#' @export
pavlab_heatmap <- function(
  mat,
  mode = c("expression", "correlation", "raw"),
  cluster_rows = FALSE,
  cluster_cols = FALSE,
  show_rownames = NULL,
  show_colnames = NULL,
  standardize_rows = NULL,
  zclip = 3,
  divergent = NULL,
  na_color = "grey80",
  palette_n = 256,
  rowname_threshold = 50,
  colname_threshold = 30,
  angle_col = 45,
  border_color = NA,
  correlation_negative_tolerance = -0.1,
  filename = NULL,
  ...
) {
  if (!requireNamespace("pheatmap", quietly = TRUE)) {
    stop("pavlab_heatmap requires the pheatmap package. install.packages('pheatmap')")
  }
  if (!is.matrix(mat)) {
    mat <- as.matrix(mat)
  }
  if (!is.numeric(mat)) {
    stop("pavlab_heatmap: matrix must be numeric (got ", class(mat)[1], ")")
  }
  mode <- match.arg(mode)

  # --- Resolve auto-defaults from mode + data shape ---

  if (is.null(standardize_rows)) {
    standardize_rows <- mode == "expression"
  }
  if (is.null(show_rownames)) {
    show_rownames <- nrow(mat) <= rowname_threshold
  }
  if (is.null(show_colnames)) {
    show_colnames <- ncol(mat) <= colname_threshold
  }

  # --- Mode-specific transforms ---

  if (mode == "correlation") {
    if (nrow(mat) == ncol(mat)) {
      diag(mat) <- NA
    } else {
      warning("pavlab_heatmap: mode='correlation' but matrix isn't square; ",
              "leaving diagonal alone")
    }
  }

  if (isTRUE(standardize_rows)) {
    mat <- t(scale(t(mat)))
    if (!is.null(zclip) && is.finite(zclip) && zclip > 0) {
      mat[!is.na(mat) & mat >  zclip] <-  zclip
      mat[!is.na(mat) & mat < -zclip] <- -zclip
    }
  }

  # --- Palette + symmetric breaks for divergent ---

  if (is.null(divergent)) {
    if (mode == "expression") {
      divergent <- TRUE   # Z-scores always span 0
    } else if (mode == "correlation") {
      # Correlation heatmaps where all values are positive (common — every
      # pair of samples / genes is at least loosely positively correlated)
      # are clearer with the sequential black-body palette. Small negative
      # noise within tolerance gets clipped to 0.
      min_val <- suppressWarnings(min(mat, na.rm = TRUE))
      if (!is.finite(min_val)) {
        divergent <- FALSE
      } else if (min_val >= correlation_negative_tolerance) {
        if (min_val < 0) {
          mat[!is.na(mat) & mat < 0] <- 0
        }
        divergent <- FALSE
      } else {
        divergent <- TRUE
      }
    } else {  # raw
      vals <- mat[is.finite(mat)]
      divergent <- (length(vals) > 0) && any(vals < 0) && any(vals > 0)
    }
  }

  if (isTRUE(divergent)) {
    palette <- divergent_palette(palette_n)
    max_abs <- suppressWarnings(max(abs(mat), na.rm = TRUE))
    if (!is.finite(max_abs) || max_abs == 0) max_abs <- 1
    breaks <- seq(-max_abs, max_abs, length.out = length(palette) + 1)
  } else {
    palette <- black_body_palette(palette_n)
    breaks <- NA   # let pheatmap auto-scale
  }

  # --- Defer to pheatmap ---

  call_args <- list(
    mat = mat,
    color = palette,
    cluster_rows = cluster_rows,
    cluster_cols = cluster_cols,
    show_rownames = show_rownames,
    show_colnames = show_colnames,
    na_col = na_color,
    angle_col = angle_col,
    border_color = border_color
  )
  if (!is.null(filename)) {
    call_args$filename <- filename
  }
  # Only pass breaks if we computed them — otherwise pheatmap should auto-choose.
  if (length(breaks) > 1 && !all(is.na(breaks))) {
    call_args$breaks <- breaks
  }

  # Merge user-provided ... arguments; user values override our defaults.
  extra <- list(...)
  for (nm in names(extra)) call_args[[nm]] <- extra[[nm]]

  do.call(pheatmap::pheatmap, call_args)
}
